Sign in Styling with FireBase🔥

oupadhyay opened the pull request:
This connects the authentication from @sanjayyepuri with the Sign-in component I had made earlier. Once merged, the page would be found at `localhost:3000/sign-in` whereas the old one is still available at `localhost:3000/signin`. 

An issue I realized at the very end is that regardless of what I log in with, the authentication accepts it. This is unlike the one from Sanjay's example which will yell at you that it does not know who you are. 

P.S. I had a NavBar on previous component design, so should I add that in as well so moving between pages is not cumbersome? The primitive NavBar's details can be found at [my repo here](https://github.com/oupadhyay/react-sign-in-up).

sanjayyepuri commented on packages/client/src/components/SignIn/index.js:
```diff
@@ -0,0 +1,113 @@
+
+    return (
+        <form>
+            { JSON.stringify(user) }
```
Remove this.

sanjayyepuri commented on packages/client/src/components/SignIn/index.js:
```diff
@@ -0,0 +1,113 @@
+                    label="Remember Me"
+                />
+            </div>
+            <style type="text/css">
+                {`
+                    .btn-coved {
+                      background-color: #00568C;
+                      color: white;
+                    }
+                `}
+            </style>
```
We shouldn't me using inline styles like this.

sanjayyepuri commented on packages/client/src/components/SignIn/index.js:
```diff
@@ -0,0 +1,113 @@
+            </style>
+            <br />
+            <Button theme="default" size="md"
+                    onClick={() => {
+                        signin(values.email, values.password)
+                    }}
```
You need to handle the promise here. Catch the failures and display a message to user. For example, if the username and password is wrong a message should be displayed notifying the user.

oupadhyay replied:
Do I need to do anything if it succeeds (i.e. a `.then()`)?

sanjayyepuri commented on packages/client/src/App.js:
```diff
@@ -1,29 +1,38 @@
-          </Route>
-          <Route path="/signin"><SignInPage/></Route>
-        </Switch>
+        <div className="auth-wrapper">
+          <div className="auth-inner">
```
Whats the purpose of these two divs. They seem to be modifying the home page styling. 
<img width="1150" alt="Screen Shot 2020-07-17 at 9 14 09 PM" src="https://user-images.githubusercontent.com/8944270/87842490-769c4080-c872-11ea-85ca-ed43c1071f3e.png">

oupadhyay replied:
Whoops. Added to specifically the SignIn component as it creates a shadow effect.

sanjayyepuri commented on packages/client/src/index.css:
```diff
@@ -13,3 +13,49 @@ code {
+}
+
+.auth-inner {
+  width: 450px;
+  margin: auto;
+  background: #ffffff;
+  box-shadow: 0 14px 80px rgba(34, 35, 58, 0.2);
+  padding: 40px 55px 45px 55px;
+  border-radius: 15px;
+  transition: all .3s;
+}
+
+.auth-wrapper .form-control:focus {
+  border-color: #00568C;
+  box-shadow: none;
+}
+
+.auth-wrapper h3 {
+  text-align: center;
+  margin: 0;
+  line-height: 1;
+  padding-bottom: 20px;
+}
+
+.custom-control-label {
+  font-weight: 400;
+}
+
+.forgot-password,
+.forgot-password a {
+  text-align: right;
+  font-size: 13px;
+  padding-top: 10px;
+  color: #7f7d7d;
+  margin: 0;
+}
+
+.forgot-password a {
+  color: #00568C;
+}
```
Can you move these styles to the component rather than making them global?

oupadhyay commented:
## Progress
Alright. I went down a rabbit hole of development, and I am now stuck. Currently, I need someone to help me find any logical errors, and why React is re-rendering so many times. 

### Logic
I have not used Formik or Yup. I know. I know. I was supposed to, but it would strip all of the component-level stylings we had done, and the only way to connect was to use a `Field` component. Most of my logic should be correct. The `serverError` is if the authentication does not succeed from `signin(email, password)` and the `error` is for any formatting issues. We should probably combine them as we don't need to provide such granularity within the sign-in component. But another set of eyes would help as the code is now very long and not very efficient (probably).

### Full Error Message
This is the full error. I have no clue how to solve this one:
```cpp
Uncaught Error: Too many re-renders. React limits the number of renders to prevent an infinite loop.
    at renderWithHooks (react-dom.development.js:14815)
    at mountIndeterminateComponent (react-dom.development.js:17482)
    at beginWork (react-dom.development.js:18596)
    at HTMLUnknownElement.callCallback (react-dom.development.js:188)
    at Object.invokeGuardedCallbackDev (react-dom.development.js:237)
    at invokeGuardedCallback (react-dom.development.js:292)
    at beginWork$1 (react-dom.development.js:23203)
    at performUnitOfWork (react-dom.development.js:22154)
    at workLoopSync (react-dom.development.js:22130)
    at performSyncWorkOnRoot (react-dom.development.js:21756)
    at react-dom.development.js:11089
    at unstable_runWithPriority (scheduler.development.js:653)
    at runWithPriority$1 (react-dom.development.js:11039)
    at flushSyncCallbackQueueImpl (react-dom.development.js:11084)
    at flushSyncCallbackQueue (react-dom.development.js:11072)
    at scheduleUpdateOnFiber (react-dom.development.js:21199)
    at dispatchAction (react-dom.development.js:15660)
    at Object.next (index.js:125)
    at subscribe.ts:104
    at subscribe.ts:233
index.js:1 The above error occurred in the <Signin> component:
    in Signin (created by Context.Consumer)
    in Route (at App.js:30)
    in Switch (at App.js:17)
    in Router (created by BrowserRouter)
    in BrowserRouter (at App.js:16)
    in AuthProvider (at App.js:14)
    in App (at src/index.js:10)
    in StrictMode (at src/index.js:9)
Consider adding an error boundary to your tree to customize error handling behavior.
Visit fb.me/react-error-boundaries to learn more about error boundaries.
```

mi-yu commented on packages/client/src/components/SignIn/index.js:
```diff
@@ -0,0 +1,173 @@
+                            id = 'email'
+                            placeholder = "Email"
+                            value = {email}
+                            onChange = {handleChange('email')}
```
you want `() => handleChange('email')` here, currently this will actually call `handleChange` every render which sets state and triggers rerender

oupadhyay replied:
So that is the one I missed. Thanks!

oupadhyay commented:
## Progress!
 - created a `ShowNotifications` component to render to allow for it to re-render.
 - styled header as `Title`
 - simplified `checkValidation()`
 - fixed `handleChange`
 - created `counter` to prevent notifications before first attempt
 - no errors

## Responses / Questions

> you want () => handleChange('email') here

Right now, putting that in `onChange` causes errors. Also, the re-rendering issue removed itself overnight so I am not sure what was the issue in the first place.

> I think alerts should be in another PR

Do you think it would be ok to leave them as is and whenever we do get around to making it, we can delete, edit, create, integrate, etc.? I just don't feel like moving all the changes into another PR and having that be a dependency for this PR...

> I have not used Formik or Yup. 

Is this ok because of the great styling hurdles it imposes?

mi-yu commented on packages/client/src/components/SignIn/index.js:
```diff
@@ -0,0 +1,186 @@
+                            endAdornment:
+                                <InputAdornment position="end">
+                                    <IconButton
+                                        onClick={() => handleClickShowPassword}
```
this should either be `onClick={() => handleClickShowPassword()}` or `onClick={handleClickShowPassword}`

mi-yu commented on packages/client/src/components/SignIn/index.js:
```diff
@@ -0,0 +1,186 @@
+                                <InputAdornment position="end">
+                                    <IconButton
+                                        onClick={() => handleClickShowPassword}
+                                        onMouseDown={() => handleMouseDownPassword}
```
same as above

mi-yu commented on packages/client/src/components/SignIn/index.js:
```diff
@@ -0,0 +1,186 @@
+
+                <div className="form-group">
+                    <Text
+                        autoFocus = {true}
```
this can just be `autoFocus`

mi-yu commented on packages/client/src/components/SignIn/index.js:
```diff
@@ -0,0 +1,186 @@
+                        placeholder = "Email"
+                        value = {email}
+                        onChange = {handleChange('email')}
+                        required = {true}
```
same as above

mi-yu commented on packages/client/src/components/SignIn/index.js:
```diff
@@ -0,0 +1,186 @@
+        return null;
+    }
+
+    // noinspection HtmlUnknownTarget
```
what is this comment for?

oupadhyay replied:
An inspection error that says that it cannot find a forgot-password page.

```
Cannot resolve file 'forgot-password'
```

mi-yu commented on packages/client/src/components/SignIn/index.js:
```diff
@@ -0,0 +1,186 @@
+        setShowPassword(!showPassword);
+    };
+    const checkValidation = () => {
+        error = !(email.length > 0 && email.includes('@') && email.includes('.') && password.length > 5);
```
Should not mutate variables like this since it's possible the component re-renders and `error` gets reset. This function should just return `true` or `false`, and can be renamed to something like `validateFields`

oupadhyay replied:
I need the `error` state to change to make sure that the notification reflects any changes to the login credentials. This was probably the easiest way to make that happen, but there is probably a better way to do this, so if there is, please let me know.

mi-yu commented on packages/client/src/components/SignIn/index.js:
```diff
@@ -0,0 +1,186 @@
+};
+
+const Signin = () => {
+    const [counter, setCounter] = useState(0);
```
nit: group and alphabetize the `useState` lines separately from `useAuth`

mi-yu commented on packages/client/src/components/SignIn/index.js:
```diff
@@ -0,0 +1,186 @@
+    }
+
+    const ShowNotifications = () => {
+        if (counter > 0) {
```
it's not clear what this counter is for, can you elaborate a bit?

oupadhyay replied:
> * created `counter` to prevent notifications before the first attempt

I need to prevent it from rendering before the user enters their credentials for the first time.

mi-yu commented on packages/client/src/components/Notification/index.js:
```diff
@@ -0,0 +1,36 @@
+ */
+
+function Notification (props) {
+    if (props.id === 'sign-in') {
```
`id` is canonically reserved for giving a unique identifier to an html element, we shouldn't use to signify a category/type since we might have multiple notifications of the same type at once on the screen.

In general it seems simpler for the consuming components to just use `UncontrolledAlert` directly

oupadhyay replied:
Done and styled.

mi-yu commented:
also not too concerned about formik/yup stuff, they can be added in another PR

johancc commented:
Seems good to me, merging to master.
